### PR TITLE
Addition of comment based help

### PR DESCRIPTION
**Pull request by Celerium**
Addresses issue #56 

Quite a bit to go over with this initial update so **ZERO rush** to get this in place but overall 99% of the changes are the addition of comment-based help to all wrapper functions with support for the `Get-Help <command name> -online` function.

Example:
`Get-Help Get-ITGlueUsers -online`

- This will take you to [https://api.itglue.com/developer/#accounts-users](https://api.itglue.com/developer/#accounts-users)

 I believe I have everything in a good initial state but with writing some much documentation it wouldn't surprise me if there are grammar or consistency issues between the wrappers. I have tested most of the functions after the modifications and I have not found any issues yet but I am only familiar with GET methods right now.

What are everyone's thoughts on the structure and or next steps?

- Is there a better way to relay\show JSON examples in the help comments?

**Unless otherwise defined all modifications are additions of comment-based help**

0. **Pester Tests**
- Updated to Pester 5
- General test improvements _(could use some more tweaks)_
- Added HelpComment.Tests to help validate help-based comments

1. **APIKey**
- Added cmdletbinding to all functions
- Fixed basic scriptAnalyzer warnings
- New Test-ITGlueAPIKey
  - Added because its helpful when needing to validate general functionality or when using RMM deployment tools

2. **BaseURI**
- Added cmdletbinding to all functions
- Fixed basic scriptAnalyzer warnings
	
3. **ModuleSettings**
- Added cmdletbinding to all functions
- Fixed basic scriptAnalyzer warnings
- New Remove-ITGlueModuleSettings
  - Nice to remove saved settings from systems when no longer needed
	
5. **ConfigurationInterfaces**
- Added cmdletbinding to New-ITGlueConfigurationInterfaces
	
6. **Configurations**
- Added cmdletbinding to New-ITGlueConfigurations

7. **ConfigurationStatues**
- Added cmdletbinding to New-ITGlueConfigurationStatuses
- Added cmdletbinding to Set-ITGlueConfigurationStatuses
	
8. **ConfigurationTypes**
  - Added cmdletbinding to New-ITGlueConfigurationTypes
  - Added cmdletbinding to Set-ITGlueConfigurationTypes
	
9. **Contacts**
- Added cmdletbinding to New-ITGlueContacts

10. **ContactTypes**
- Added cmdletbinding to New-ITGlueContactTypes
- Added cmdletbinding to Set-ITGlueContactTypes
	
11. **FlexibleAssetFields**
- Added cmdletbinding to New-ITGlueFlexibleAssetFields
- _Might need to add in mandatory parameters to Get-ITGlueFlexibleAssetFields (Can come later)_
	
12. **FlexibleAssetTypes**
- Added cmdletbinding to New-ITGlueFlexibleAssetTypes
- Added cmdletbinding to Set-ITGlueFlexibleAssetTypes
	
13. **FlexibleAssets**
- _Might need to add mandatory parameter's to Get-ITGlueFlexibleAssets (Can come later)_

14. **Locations**
- Added cmdletbinding to New-ITGlueLocations
- _Should the "Remove-ITGlueLocations" data param need an update paramSet?_
	
15. **Manufacturers**
- Added cmdletbinding to New-ITGlueManufacturers
- Added cmdletbinding to Set-ITGlueManufacturers
	
16. **Models**
- Added cmdletbinding to New-ITGlueModels
	
17. **Organizations**
- Added cmdletbinding to New-ITGlueOrganizations
	
18. **OrganizationStatuses**
- Added cmdletbinding to New-ITGlueOrganizationStatuses
- Added cmdletbinding to Set-ITGlueOrganizationStatuses
	
19. **OrganizationTypes**
- Added cmdletbinding to New-ITGlueOrganizationTypes
- Added cmdletbinding to Set-ITGlueOrganizationTypes
	
20. **PasswordCategories**
- Added cmdletbinding to New-ITGluePasswordCategories
- Added cmdletbinding to Set-ITGluePasswordCategories
	
21. **Passwords**
- Added cmdletbinding to New-ITGluePasswords

22. **Users**
- Added cmdletbinding to Set-ITGlueUsers

### CODE REVIEW

File: ITGlueAPI/Internal/ModuleSettings.ps1
Comment (CalebAlbers) [inline ITGlueAPI/Internal/ModuleSettings.ps1:68]: The rest of the module tends to use `snake_case` instead of `PascalCase` - I recognize that it goes against Microsoft's best practices, but I think consistency within the module should take precedence. 

Adding in parameter aliases so that new people can use the `PascalCase` scheme while existing scripts fall back to `snake_case` is a good topic for a future PR, though - I'll file an Issue for it.

**Review by CalebAlbers — COMMENTED** [review]
Phenomenal work, @Celerium! This undoubtedly took a ton of time. I haven't gone through the entire PR yet, but a few things spring to mind:

> with support for the Get-Help <command name> -online function.
>
>`Get-Help Get-ITGlueUsers -online`
> This will take you to https://api.itglue.com/developer/#accounts-users

I think this is a fantastic add - definitely something I wish I had when originally using the module 🙂 

> What are everyone's thoughts on the structure and or next steps?

I left a few comments, but the biggest thing that calls out to me would be considering inlining the parameter descriptions. I'd love to have some dialogue on that with you.

> Is there a better way to relay\show JSON examples in the help comments?
I'm not sure that there really is a great option there. This goes into a potentially larger discussion on how to best create the PSObjects that these functions use (maybe a generator of some kind, or use of object types?), but I think leaving the documentation in the state it is in currently is a great option, and we can build on it over time.

File: ITGlueAPI/Resources/Configurations.ps1
Comment (Celerium) [reply ITGlueAPI/Resources/Configurations.ps1:125]: Interesting, I was today years old when I found out this was possible. 

Overall I feel that it does have its use with small modules, functions, and really just any PoSH script that doesn't have a lot of parameters\parameterSets.  In fact, I might actually use this for some of the smaller helper functions I've written. However, when it comes to larger modules like this one it feels more natural\cleaner to keep the help section separate from the parameter section.

I did some quick looking at some of the more popular modules on GitHub and even some of the MS ones and I haven't found any other PoSH module that implements in-line help yet.

You do make a good point about how this could be a good way of encouraging others to document their parameters as they add them. Something to note though is how the parameter section might look if some of the descriptions run a little longer or are multi-lined. If everything is kept super short and to a single line then it feels like it wouldn't cause any issues.

- When it comes to `Get-Help` they both display the descriptions to the user in the same way.
- With either option of deploying about_help, the Pester `HelpComment` test would help validate any missing descriptions. 

I drafted a quick example of what it could look like below. 

```powershell
function Test-InLineHelp {

<#
.SYNOPSIS
    .
.DESCRIPTION
    .
.PARAMETER Path
    The path to the .
.PARAMETER LiteralPath
    Specifies a literal path to one or more locations.
#>

    Param(
        [String]$Path = "Help",

        # This is a simple sentence about this thing but here is also an FYI
        # You should also do this other thing because stuff
        [String]$SimpleParam = "Help",

        <#
            This is my NEW target
            This is another line to the target
            THIS WORKS to New Targets
        #>
        [String]$NewTargets = "Help",

        [String]$TestTargets = "Help",
        <#
            This is my TESTtarget
            This is another line to the target
            THIS WILL GET ASSOCIATED TO AnotherTarget
        #>

        [String]$AnotherTarget = "Help"
            #This is my ANOTHER target
            #This is another line to the target
            #TEST Target and Another Target help get combined

    )

}
```

```
NAME
    Test-InLineHelp
    
SYNOPSIS
    .
    
    
SYNTAX
    Test-InLineHelp [[-Path] <String>] [[-SimpleParam] <String>] [[-NewTargets] <String>] [[-TestTargets] <String>] [[-AnotherTarget] <String>] [<CommonParameters>]
    
    
DESCRIPTION
    .


PARAMETERS
    -Path <String>
        The path to the .

        Required?                    false
        Position?                    1
        Default value                Help
        Accept pipeline input?       false
        Accept wildcard characters?  false

    -SimpleParam <String>
        This is a simple sentence about this thing but here is also an FYI
        You should also do this other thing because stuff

        Required?                    false
        Position?                    2
        Default value                Help
        Accept pipeline input?       false
        Accept wildcard characters?  false

    -NewTargets <String>
        This is my NEW target
        This is another line to the target
        THIS WORKS to New Targets

        Required?                    false
        Position?                    3
        Default value                Help
        Accept pipeline input?       false
        Accept wildcard characters?  false

    -TestTargets <String>

        Required?                    false
        Position?                    4
        Default value                Help
        Accept pipeline input?       false
        Accept wildcard characters?  false

    -AnotherTarget <String>
        This is my TESTtarget
        This is another line to the target
        THIS WILL GET ASSOCIATED TO AnotherTarget

        This is my ANOTHER target
        This is another line to the target
        TEST Target and Another Target help get combined

        Required?                    false
        Position?                    5
        Default value                Help
        Accept pipeline input?       false
        Accept wildcard characters?  false

    <CommonParameters>
        This cmdlet supports the common parameters: Verbose, Debug,
        ErrorAction, ErrorVariable, WarningAction, WarningVariable,
        OutBuffer, PipelineVariable, and OutVariable. For more information, see
        about_CommonParameters (https:/go.microsoft.com/fwlink/?LinkID=113216).

INPUTS
OUTPUTS
RELATED LINKS
```

File: ITGlueAPI/Resources/Configurations.ps1
Comment (Celerium) [reply ITGlueAPI/Resources/Configurations.ps1:628]: I plan on taking some time this week\weekend to correct some grammar and make the SYNOPSIS, DESCRIPTION, and common PARAMETER names more consistent between the functions.

File: ITGlueAPI/Tests/HelpComment.Tests.ps1
Comment (Celerium) [reply ITGlueAPI/Tests/HelpComment.Tests.ps1:115]: Thank you, I pieced that together a while ago from a few different sources when I was updating documentation in bulk for older scripts, and it has really helped speed up help validations.

File: ITGlueAPI/Internal/ModuleSettings.ps1
Comment (Celerium) [reply ITGlueAPI/Internal/ModuleSettings.ps1:68]: I can get that switched over for consistency. How does the following look?

- itglue_api_conf_path
- itglue_api_conf_file
- open_conf_Path    (renamed config to conf)

To be honest I didn't even think about it with this section as I copied what was there previously and moved them into parameters.

```powershell
# Locations of settings path and file
$ITGlueAPIConfPath = "$($env:USERPROFILE)\ITGlueAPI"
$ITGlueAPIConfFile = "config.psd1"
```

Completely agree with

> but I think consistency within the module should take precedence.

File: ITGlueAPI/Resources/Configurations.ps1
Comment (CalebAlbers) [reply ITGlueAPI/Resources/Configurations.ps1:125]: Seeing it written out like that really makes it easy to lose sight of the parameter section. I'm down to roll with having everything in one place instead of splitting it up!

File: ITGlueAPI/Internal/ModuleSettings.ps1
Comment (CalebAlbers) [reply ITGlueAPI/Internal/ModuleSettings.ps1:68]: Those look good! Figures it'd be a holdover from some code I wrote years ago 😆

**Comment by Celerium** [comment]
With all the new development going on when would be a good time to merge these changes? I was thinking once #157 is completed.

This would give me or others time to review any grammar or consistency issues with help descriptions between the functions as well as give a good jumping-off point for further module development.

File: ITGlueAPI/ITGlueAPI.psd1
Comment (Celerium) [inline ITGlueAPI/ITGlueAPI.psd1:110]: Once #157 is complete and if this is something that has value to add to the module I can migrate the `Invoke` section to the new internal API call function(s).

**Comment by Celerium** [comment]
I've updated various, grammar, spelling, and word consistency for each of the functions and feel like this is a good starting point for the modules about_help.

I can switch this to ready for review so it can be merged now or after #157.

**Comment by adrianwells** [comment]
@Celerium this is an impressive amount of work as @CalebAlbers noted. It seems it should be merged soon. @davidhaymond thoughts?

**Comment by Celerium** [comment]
> @Celerium this is an impressive amount of work as @CalebAlbers noted. It seems it should be merged soon. @davidhaymond thoughts?

@adrianwells thank you, something to note as it's been a while, so it would be a good idea to do a once-over of this addition as well as make sure any of the new functions include comment-based help.

**Comment by davidhaymond** [comment]
> @Celerium this is an impressive amount of work as @CalebAlbers noted. It seems it should be merged soon. @davidhaymond thoughts?

I would love to see this merged and included in v2.3.0. I'll see if I can find time this week to review it.

**Review by davidhaymond — CHANGES_REQUESTED** [review]
I've done a quick review and made a few suggestions. Amazing work, @Celerium!

File: ITGlueAPI/Internal/APIKey.ps1
Comment (Celerium) [reply ITGlueAPI/Internal/APIKey.ps1:197]: That is a good question, the reason I used `Invoke-WebRequest` instead of `Invoke-RestMethod` here was because `Invoke-WebRequest` returns header information. I'll do some testing this weekend to see what I can do. 

As a thought what if the `Invoke-ITGlueRequest` is used to test first and if it fails then the `Invoke-WebRequest` method is called to get the failure details?

I also wouldn't be offended if we didn't include this, it is just a nice to have IMO.

Comment (davidhaymond) [reply ITGlueAPI/Internal/APIKey.ps1:197]: It's fine to keep it like this, since there's a good reason for doing so. Thanks for pointing that out!

Comment (Celerium) [reply ITGlueAPI/Internal/APIKey.ps1:197]: Sounds good to me, is there anything else I should clean up with this pull request? I believe the remaining conflicts are just line breaks.

Comment (davidhaymond) [reply ITGlueAPI/Internal/APIKey.ps1:197]: I can't think of anything, other than fixing the remaining conflicts. @adrianwells, could you do a "squash and merge" when you merge this PR? There's 11 commits now and it's a lot of clutter that probably shouldn't be in the mainline Git history.

Comment (adrianwells) [reply ITGlueAPI/Internal/APIKey.ps1:197]: @Celerium Thank you. It appears more than line breaks creating a conflict in OrganizationTypes.ps1 (the other 2 files, ITGlueAPI.psm and PasswordCategories.ps1 seem to only line break issues). You able to review and update? @davidhaymond I think is close to ready.

Comment (Celerium) [reply ITGlueAPI/Internal/APIKey.ps1:197]: @adrianwells, I could be wrong but the PasswordCategories.ps1 conflict appears to be the help section.

**Comment by davidhaymond** [comment]
@Celerium This PR seems ready to me. Could you rebase your branch on top of the current master branch and squash all the commits? I'd like to try to keep the Git history relatively clean.

It looks like you've allowed maintainers to edit the PR, so if you like, I can rebase and merge for you.

**Comment by Celerium** [comment]
> @Celerium This PR seems ready to me. Could you rebase your branch on top of the current master branch and squash all the commits? I'd like to try to keep the Git history relatively clean.
> 
> It looks like you've allowed maintainers to edit the PR, so if you like, I can rebase and merge for you.

@davidhaymond, if you could that would be wonderful. Thank you

**Comment by davidhaymond** [comment]
I've cleaned up this PR and merged into master! Thanks again for your incredible contribution, @Celerium!

**Pull request merged**
